src/Features/Core/Diagnostics/DiagnosticData.cs

```diff
@@ -284,6 +284,8 @@ private static void SwapIfNeeded(ref LinePosition startLinePosition, ref LinePos
 
         public static DiagnosticData Create(Workspace workspace, Diagnostic diagnostic)
         {
+            Contract.Requires(diagnostic.Location == null || !diagnostic.Location.IsInSource);
```

Comment:
Then, we probably want to seal Diagnostic and use different pattern then what we are doing now.


-          heejae

From: Manish Vasani [mailto:notifications@github.com]
Sent: Thursday, March 5, 2015 1:46 PM
To: dotnet/roslyn
Cc: HeeJae Chang
Subject: Re: [roslyn] support compilation end action in v1 diagnostic engine (#1047)


In src/Features/Core/Diagnostics/DiagnosticData.cs<https://github.com/dotnet/roslyn/pull/1047#discussion_r25904742>:

> @@ -284,6 +284,8 @@ private static void SwapIfNeeded(ref LinePosition startLinePosition, ref LinePos

>

>          public static DiagnosticData Create(Workspace workspace, Diagnostic diagnostic)

>          {

> +            Contract.Requires(diagnostic.Location == null || !diagnostic.Location.IsInSource);

Diagnostic has internal abstract methods so all implementations should be internal. I don't think we ever want to allow public implementations of Diagnostic, that will probably cause lots of other unwanted issues. We can probably add an assert in our ReportDiagnostic methods that it is of one of our internal Diagnostic types.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/roslyn/pull/1047/files#r25904742>.